keps/sig-autoscaling/draft-20180419-enhance-hpa-metrics-specificity.md

```diff
@@ -0,0 +1,247 @@
+
+```
+queue_length{statefulset="foo",pod="foo-1",queue="some-jobs"}
+queue_length{statefulset="foo",pod="foo-1",queue="other-jobs"}
+```
+
+Suppose they have a pool of works that they wish to scale for each queue.
+In the current-day HPA, it's non-trivial to allow selecting the metric for
+a specific queue.  Current suggestions are metric-backend-specific (for
+instance, you could create a Prometheus recording rule to relabel or
+rename the metric), and often involve making external changes to the
+metrics pipeline.
+
+With the addition of the metrics label selector, users could simply select
+the queue using the label selector:
+
+```yaml
+- type: Object
+  object:
+    target:
+      kind: StatefulSet
+      apiVersion: apps/v1
+      name: foo
+    targetValue: 2
+    metric:
+      name: queue_length
+      selector: "queue=some-jobs"
+```
+
+Similarly, in discussions of scaling on queues, being able to divide
+a target backlog length by the number of available pods is often useful --
+for instance, a backlog length of 3 might be acceptable if there are three
+pods processing items, but not if there is only one.
+
+### Goals
+
+- The autoscaling/v2 API is updated with the additional fields
+  described below.
+- A corresponding change is made to the custom metrics API to support the
+  additional label selector.
+- The testing adapter is updated to support these changes (for e2e
+  purposes).
+
+### Non-Goals
+
+It is outside of the purview of the KEP to ensure that current custom
+metrics adapters support the new changes -- this is up to those adapters
+maintainers.
+
+## Proposal
+
+The autoscaling/v2 API will be updated the following way:
+
+```go
+type ObjectMetricSource struct {
+	Target CrossVersionObjectReference
+
+	// +optional
+	TargetValue *resource.Quantity
+	// +optional
+	TargetAverageValue *resource.Quantity
+
+	Metric MetricIdent
+}
+
+type PodsMetricSource struct {
+	TargetAverageValue resource.Quantity
+	Metric MetricIdent
+}
+
+type ExternalMetricSouce struct {
+	Metric MetricIdent
+
+	// +optional
+	TargetValue *resource.Quantity
+	// +optional
+	TargetAverageValue *resource.Quantity
+}
+
+type MetricIdent struct {
+	// name is the name of the given metric
+	Name string
+	// selector is the selector for the given metric
+	// it is the string-encoded form of a standard kubernetes label selector
+	// +optional
+	Selector string
+}
+
+// and similarly for the statuses
+```
+
+Notice that the `metricName` field is replaced with a new `metric` field,
+which encapsulates both the metric name, and an optional label selector,
+which takes the form of a standard kubernetes label selector.  The
+`External` source is updated slightly to match the new form of the `Pods`
+and `Object` sources.
+
+These changes necessitate a second beta of `autoscaling/v2`:
+`autoscaling/v2beta2`.
+
+Similarly, corresponding changes need to be made to the custom metrics
+API:
+
+```go
+type MetricValue struct {
```

Comment:
In external metrics API we return a list of values, one for each time series. I think we need to do that in CM API as well or we need to say that it's adapters job to aggregate all time series into a single value.

Reply:
Yeah, the intention was that it's the adapter's job to aggregate, if necessary.  That's the approach we took before the specificity, and I missed that the external metrics API doesn't do that (I'd prefer if it did).  I'll note that explicitly